Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,336 @@
+    [Dependency] private readonly RoundEndSystem _roundEndSystem = default!;
+    [Dependency] private readonly SharedAudioSystem _audioSystem = default!;
+    [Dependency] private readonly MindSystem _mindSystem = default!;
+    [Dependency] private readonly SharedStunSystem _sharedStun = default!;
+
+    private ISawmill _sawmill = default!;
+
+    private int MinPlayers => _cfg.GetCVar(CCVars.RevolutionaryMinPlayers);
+    private int MaxRevHeads => _cfg.GetCVar(CCVars.RevolutionaryMaxHeadRevs);
+    private int HeadRevsPerPlayer => _cfg.GetCVar(CCVars.RevolutionaryPlayersPerHeadRev);
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        _sawmill = Logger.GetSawmill("preset");
+        SubscribeLocalEvent<RoundStartAttemptEvent>(OnStartAttempt);
+        SubscribeLocalEvent<RulePlayerJobsAssignedEvent>(OnPlayerJobAssigned);
+        SubscribeLocalEvent<PlayerSpawnCompleteEvent>(OnPlayerSpawned);
+        SubscribeLocalEvent<MobStateChangedEvent>(OnMobStateChanged);
+        SubscribeLocalEvent<RoundEndTextAppendEvent>(OnRoundEndText);
+    }
+
+    private void OnRoundEndText(RoundEndTextAppendEvent ev)
+    {
+        foreach (var headrev in EntityQuery<RevolutionaryRuleComponent>())
+        {
+            if (headrev.HeadsDied && !headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-won"));
+            }
+            if (!headrev.HeadsDied && headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-lost"));
+            }
+            if (!headrev.HeadsDied && !headrev.RevsLost)
+            {
+                ev.AddLine(Loc.GetString("rev-stalemate"));
+            }
+            ev.AddLine(Loc.GetString("head-rev-initial-count", ("initialCount", headrev.HeadRevs.Count)));
+            foreach (var player in headrev.HeadRevs)
+            {
+                ev.AddLine(Loc.GetString("head-rev-initial",
+                    ("name", player.Key),
+                    ("username", player.Value)));
+            }
+        }
+    }
+    private void OnStartAttempt(RoundStartAttemptEvent ev)
+    {
+        var query = EntityQueryEnumerator<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+            if (!ev.Forced || ev.Players.Length < MinPlayers)
+            {
+                _chatManager.SendAdminAnnouncement(Loc.GetString("rev-not-enough-ready-players",
+                    ("readyPlayersCount", ev.Players.Length),
+                    ("minimumPlayers", MinPlayers)));
+                ev.Cancel();
+                continue;
+            }
+            if (ev.Players.Length == 0)
+            {
+                _chatManager.DispatchServerAnnouncement(Loc.GetString("rev-no-one-ready"));
+                ev.Cancel();
+            }
+        }
+    }
```

Comment:
This uses the struct enumerator not the IEnumerable version like above which tells me you copied it so please move it to a generic method instead of copy-pasting code.